### PR TITLE
demux_lavf: replaygain export simplification

### PR DESCRIPTION
**Pull request by jeeb**
1. Utilize `av_stream_get_side_data` instead of looping over all side data entries. This simplifies the code.
    
    In theory if you had multiple entries in a single stream for the same side data type, the looping could be required, but as far as I can tell 41ba6dad721e618c0e54315bd72ad28afbd7b622 makes no mention of this being an issue, and I don't think the official side data helper functions ever made this possible? Thus this probably was just missing the required helper functionality or it wasn't noticed back in 2014?
2. Simplify the function by utilizing early exit and always setting the album data to track data, if track data is available. This lessens the depth of required if/else structures and makes sure we do not actually export a `replaygain_data` structure internally if there was no useful data in the side data (at least one of the sides - track or album should have non-default values).

This can still leave us with a `replaygain_data` struct with track side being set to defaults, but that is the current state of affairs as well.

### CODE REVIEW

**Comment by kasper93** [comment]
Could you rebase? Thanks.

File: demux/demux_lavf.c [hunk 1]
```diff
@@ -603,48 +603,50 @@ static void select_tracks(struct demuxer *demuxer, int start)
...
-
-        if (src_sd->type != AV_PKT_DATA_REPLAYGAIN)
-            continue;
+    struct replaygain_data *rgain = NULL;
+    const AVPacketSideData *side_data =
+        av_packet_side_data_get(st->codecpar->coded_side_data,
+                                st->codecpar->nb_coded_side_data,
+                                AV_PKT_DATA_REPLAYGAIN);
+    if (!side_data || !side_data->data)
+        return;
```
Comment (kasper93) [inline demux/demux_lavf.c:612]:
``` suggestion
    const AVReplayGain *av_rgain  = (void *)mp_av_stream_get_side_data(st, AV_PKT_DATA_REPLAYGAIN);
    if (!av_rgain)
        return;
```

File: demux/demux_lavf.c [hunk 2]
```diff
@@ -603,48 +603,50 @@ static void select_tracks(struct demuxer *demuxer, int start)
...
-                rgain->album_gain = av_rgain->album_gain / 100000.0f;
-                rgain->album_peak = av_rgain->album_peak / 100000.0f;
-            } else {
-                // Album gain is undefined; fall back to track gain.
-                rgain->album_gain = rgain->track_gain;
-                rgain->album_peak = rgain->track_peak;
-            }
-        }
+    const AVReplayGain *av_rgain =
+        (const AVReplayGain*)side_data->data;
```
Comment (kasper93) [inline demux/demux_lavf.c:615]:
```suggestion
```

File: demux/demux_lavf.c [hunk 3]
```diff
@@ -603,48 +603,50 @@ static void select_tracks(struct demuxer *demuxer, int start)
...
+
+    bool track_data_available =
+        (av_rgain->track_gain != INT32_MIN && av_rgain->track_peak != 0);
```
Comment (kasper93) [inline demux/demux_lavf.c:618]:
```suggestion
        av_rgain->track_gain != INT32_MIN && av_rgain->track_peak != 0;
```
style

File: demux/demux_lavf.c [hunk 4]
```diff
@@ -603,48 +603,50 @@ static void select_tracks(struct demuxer *demuxer, int start)
...
+        (av_rgain->album_gain != INT32_MIN && av_rgain->album_peak != 0);
 
-        // This must be run only before the stream was added, otherwise there
-        // will be race conditions with accesses from the user thread.
-        assert(!sh->ds);
-        sh->codec->replaygain_data = rgain;
+    if (!track_data_available && !album_data_available)
+        return;
+
+    rgain    = talloc_ptrtype(demuxer, rgain);
```
Comment (kasper93) [inline demux/demux_lavf.c:625]:
``` suggestion
    struct replaygain_data *rgain = talloc_ptrtype(demuxer, rgain);
```

File: demux/demux_lavf.c [hunk 5]
```diff
@@ -603,48 +603,50 @@ static void select_tracks(struct demuxer *demuxer, int start)
...
-                rgain->album_peak = rgain->track_peak;
-            }
-        }
+    const AVReplayGain *av_rgain =
+        (const AVReplayGain*)side_data->data;
+
+    bool track_data_available =
+        (av_rgain->track_gain != INT32_MIN && av_rgain->track_peak != 0);
+    bool album_data_available =
+        (av_rgain->album_gain != INT32_MIN && av_rgain->album_peak != 0);
```
Comment (kasper93) [inline demux/demux_lavf.c:620]: Is it possible to have gain without peak value?

Comment (jeeb) [reply demux/demux_lavf.c:620]: This is just the same check being moved from inside the ifs to variables, so unfortunately I don't have the answer to that.

For better or worse the behavior check-wise should at least be the same.

**Review by kasper93 — APPROVED** [review]

**Pull request merged**